In [1]:
import pandas as pd

**Phase 1: Explore the Data**  
Objectives:  
* Understand the purpose of each dataset.  
* Inspect column types, sizes, and example rows.  

In [13]:
# Read the fullsample dataset into jobs dataframe
jobs = pd.read_csv("../data/fullsample.csv")
jobs.head()

JOBID      STATE                BEGIN                  END    REQMEM  \
0    30616928    RUNNING  2021-07-31T22:15:00              Unknown    2048Mn   
1    30853133  COMPLETED  2021-08-06T11:36:09  2021-09-05T11:36:32  262144Mn   
2    30858137  COMPLETED  2021-08-06T19:04:39  2021-09-05T19:04:53  204800Mn   
3    30935078  COMPLETED  2021-08-09T16:52:51  2021-09-07T20:52:55   65536Mn   
4  31364111_2  COMPLETED  2021-08-17T07:45:07  2021-09-10T16:45:24   16384Mn   

     USEDMEM      REQTIME     USEDTIME  NODES  CPUS     PARTITION EXITCODE  
0          0     10:04:00  67-22:14:22      1     1    production      0:0  
1  20604.62M  30-00:00:00  30-00:00:23      1     1  cgw-platypus      0:0  
2  57553.77M  30-00:00:00  30-00:00:14      1    32     cgw-tbi01      0:0  
3  20577.96M  29-04:00:00  29-04:00:04      1     8  cgw-platypus      0:0  
4   9733.43M  24-09:00:00  24-09:00:17      1     1    production      0:0

In [16]:
jobs.shape

(7395885, 12)

* jobs dataframe contains 7395885 rows and 12 columns

* The fullsample dataset contains job records, with one row per job.

* Each job gets a unique ID, contained in the **JOBID** column.

* Some jobs can be submitted as arrays of similar jobs. These are listed with an underscore in the JOBID, where the number after the   underscore indicates the tasknumber. For example. JOBID 31781951 was an array job with 10 parts.

In [10]:
jobs.info(show_counts = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7395885 entries, 0 to 7395884
Data columns (total 12 columns):
 #   Column     Non-Null Count    Dtype 
---  ------     --------------    ----- 
 0   JOBID      7395885 non-null  object
 1   STATE      7395885 non-null  object
 2   BEGIN      7395885 non-null  object
 3   END        7395885 non-null  object
 4   REQMEM     7395885 non-null  object
 5   USEDMEM    7395885 non-null  object
 6   REQTIME    7395885 non-null  object
 7   USEDTIME   7395885 non-null  object
 8   NODES      7395885 non-null  int64 
 9   CPUS       7395885 non-null  int64 
 10  PARTITION  7395885 non-null  object
 11  EXITCODE   7395885 non-null  object
dtypes: int64(2), object(10)
memory usage: 677.1+ MB


* All the columns in the jobs dataframe contain non null values

In [48]:
#convert the BEGIN and END columns to date time format
jobs['BEGIN'] = pd.to_datetime(jobs['BEGIN'],
                                           format = "%Y-%m-%dT%H:%M:%S", errors='coerce')

jobs['END'] = pd.to_datetime(jobs['END'],
                                           format = "%Y-%m-%dT%H:%M:%S", errors='coerce')


In [41]:
jobs.describe()

NODES          CPUS
count  7.395885e+06  7.395885e+06
mean   1.005995e+00  1.861186e+00
std    4.292460e-01  3.158301e+00
min    1.000000e+00  0.000000e+00
25%    1.000000e+00  1.000000e+00
50%    1.000000e+00  1.000000e+00
75%    1.000000e+00  1.000000e+00
max    5.000000e+02  1.280000e+02

In [51]:
jobs.sample(5)

JOBID      STATE               BEGIN                 END  \
775511     32680158_29  COMPLETED 2021-09-20 17:27:48 2021-09-20 17:29:36   
2024499  31189316_5379  COMPLETED 2021-08-12 21:47:23 2021-08-12 21:55:24   
5495724   26784999_135  COMPLETED 2021-01-14 17:05:59 2021-01-14 20:01:37   
506490     32427395_38  COMPLETED 2021-09-14 13:52:44 2021-09-14 14:11:27   
7339981   25402574_124  COMPLETED 2020-10-28 00:46:48 2020-10-28 00:48:23   

          REQMEM   USEDMEM     REQTIME  USEDTIME  NODES  CPUS   PARTITION  \
775511   16384Mn   385.15M  2-02:00:00  00:01:48      1     1  production   
2024499   4096Mn   178.04M    08:00:00  00:08:01      1     1  production   
5495724   4096Mc  1442.04M  2-00:00:00  02:55:38      1     1  production   
506490    4096Mc    79.24M    10:00:00  00:18:43      1     1  production   
7339981  24576Mn         0    00:30:00  00:01:35      1     2  production   

        EXITCODE  
775511       0:0  
2024499      0:0  
5495724      0:0  
506490       0:0  
7339981      0:0

The __slurm_wrapper_ce5.log__ and __slurm_wrapper_ce6.log__ files contain logs of jobs submitted from the Open Science Grid.

In [25]:
# Read the slurm_wrapper_ce5.log dataset into jobs ce5

ce5 = pd.read_csv('../data/slurm_wrapper_ce5.log',
                  header=None,
                  delimiter=' - ',
                  engine='python')

ce5.head()

0       1        2                         3  \
0  2020-10-16 08:15:39.278699  user 0  retry 0  time 0.07347559928894043   
1  2020-10-16 08:18:08.313309  user 0  retry 0  time 0.18363237380981445   
2  2020-10-16 08:22:48.128689  user 0  retry 0  time 0.07547116279602051   
3  2020-10-16 08:25:13.257408  user 0  retry 0  time 0.09484362602233887   
4  2020-10-16 08:31:01.460723  user 0  retry 0  time 0.07498788833618164   

              4                                                  5  
0  returncode 0  command ['/usr/bin/sacct', '-u', 'appelte1', '...  
1  returncode 0  command ['/usr/bin/sacct', '-u', 'appelte1', '...  
2  returncode 0  command ['/usr/bin/sacct', '-u', 'appelte1', '...  
3  returncode 0  command ['/usr/bin/sacct', '-u', 'appelte1', '...  
4  returncode 0  command ['/usr/bin/sacct', '-u', 'appelte1', '...

In [30]:
ce5.shape

(4770893, 6)

* ce5 dataframe contains 4770893 rows and 6 columns

In [31]:
ce5.info(show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4770893 entries, 0 to 4770892
Data columns (total 6 columns):
 #   Column  Non-Null Count    Dtype 
---  ------  --------------    ----- 
 0   0       4770893 non-null  object
 1   1       4770893 non-null  object
 2   2       4770893 non-null  object
 3   3       4770893 non-null  object
 4   4       4770893 non-null  object
 5   5       4770893 non-null  object
dtypes: object(6)
memory usage: 218.4+ MB


In [39]:
# Rename the column names of ce5 dataframe

new_column_names = ['date', 'user', 'retry' , 'time', 'return_code', 'command']
ce5.columns = new_column_names
ce5.head()

date    user    retry                      time  \
0  2020-10-16 08:15:39.278699  user 0  retry 0  time 0.07347559928894043   
1  2020-10-16 08:18:08.313309  user 0  retry 0  time 0.18363237380981445   
2  2020-10-16 08:22:48.128689  user 0  retry 0  time 0.07547116279602051   
3  2020-10-16 08:25:13.257408  user 0  retry 0  time 0.09484362602233887   
4  2020-10-16 08:31:01.460723  user 0  retry 0  time 0.07498788833618164   

    return_code                                            command  
0  returncode 0  command ['/usr/bin/sacct', '-u', 'appelte1', '...  
1  returncode 0  command ['/usr/bin/sacct', '-u', 'appelte1', '...  
2  returncode 0  command ['/usr/bin/sacct', '-u', 'appelte1', '...  
3  returncode 0  command ['/usr/bin/sacct', '-u', 'appelte1', '...  
4  returncode 0  command ['/usr/bin/sacct', '-u', 'appelte1', '...

In [49]:
#convert the date column datatype  to date time format

ce5['date'] = pd.to_datetime(ce5['date'],
                                           format = "%Y-%m-%d %H:%M:%S", errors='coerce')


In [53]:
ce5.sample(5)

date       user    retry                      time   return_code  \
3434081  NaT  user 9201  retry 0  time 0.04343152046203613  returncode 0   
2485050  NaT  user 9201  retry 0    time 6.328029155731201  returncode 0   
129808   NaT  user 9201  retry 0  time 0.04030132293701172  returncode 0   
219307   NaT  user 9219  retry 0   time 1.7425882816314697  returncode 0   
33323    NaT  user 9202  retry 0  time 0.22401785850524902  returncode 0   

                                                   command  
3434081  command ['/usr/bin/sbatch', '/tmp/condor_g_scr...  
2485050       command ['/usr/bin/scontrol', 'show', 'job']  
129808   command ['/usr/bin/sbatch', '/tmp/condor_g_scr...  
219307        command ['/usr/bin/scontrol', 'show', 'job']  
33323    command ['/usr/bin/sbatch', '/tmp/condor_g_scr...

In [18]:
# Read the slurm_wrapper_ce6.log dataset into jobs ce6

ce6 = pd.read_csv('../data/slurm_wrapper_ce6.log',
                  header=None,
                  delimiter=' - ',
                  engine='python')

ce6.head()

0          1        2                          3  \
0  2020-10-16 10:37:44.163454  user 9202  retry 0   time 0.08495402336120605   
1  2020-10-16 10:37:44.206654  user 9202  retry 0   time 0.08943057060241699   
2  2020-10-16 10:37:44.218760  user 9202  retry 0   time 0.05928945541381836   
3  2020-10-16 10:37:44.256403  user 9202  retry 0  time 0.038695573806762695   
4  2020-10-16 10:37:44.611603  user 9202  retry 0   time 0.03343677520751953   

              4                                                  5  
0  returncode 0  command ['/usr/bin/scontrol', 'show', 'job', '...  
1  returncode 0  command ['/usr/bin/scontrol', 'show', 'job', '...  
2  returncode 0  command ['/usr/bin/scontrol', 'show', 'job', '...  
3  returncode 0  command ['/usr/bin/scontrol', 'show', 'job', '...  
4  returncode 0  command ['/usr/bin/scontrol', 'show', 'job', '...

In [20]:
ce6.shape

(4776520, 6)

ce6 dataframe contains 4776520 rows and 6 columns

In [24]:
ce6.info(show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4776520 entries, 0 to 4776519
Data columns (total 6 columns):
 #   Column  Non-Null Count    Dtype 
---  ------  --------------    ----- 
 0   0       4776520 non-null  object
 1   1       4776520 non-null  object
 2   2       4776520 non-null  object
 3   3       4776520 non-null  object
 4   4       4776520 non-null  object
 5   5       4776520 non-null  object
dtypes: object(6)
memory usage: 218.7+ MB


In [40]:
#Rename the column names

new_column_names = ['date', 'user', 'retry' , 'time', 'return_code', 'command']
ce6.columns = new_column_names
ce6.head()

date       user    retry                       time  \
0  2020-10-16 10:37:44.163454  user 9202  retry 0   time 0.08495402336120605   
1  2020-10-16 10:37:44.206654  user 9202  retry 0   time 0.08943057060241699   
2  2020-10-16 10:37:44.218760  user 9202  retry 0   time 0.05928945541381836   
3  2020-10-16 10:37:44.256403  user 9202  retry 0  time 0.038695573806762695   
4  2020-10-16 10:37:44.611603  user 9202  retry 0   time 0.03343677520751953   

    return_code                                            command  
0  returncode 0  command ['/usr/bin/scontrol', 'show', 'job', '...  
1  returncode 0  command ['/usr/bin/scontrol', 'show', 'job', '...  
2  returncode 0  command ['/usr/bin/scontrol', 'show', 'job', '...  
3  returncode 0  command ['/usr/bin/scontrol', 'show', 'job', '...  
4  returncode 0  command ['/usr/bin/scontrol', 'show', 'job', '...

In [50]:
#convert the date column datatype  to date time format

ce6['date'] = pd.to_datetime(ce6['date'],
                                           format = "%Y-%m-%d %H:%M:%S", errors='coerce')

In [54]:
ce5.sample(5)

date       user    retry                       time   return_code  \
1633408  NaT  user 9201  retry 0  time 0.031569719314575195  returncode 0   
1941388  NaT  user 9201  retry 0    time 0.6722025871276855  returncode 0   
4734953  NaT  user 9201  retry 0   time 0.08736014366149902  returncode 0   
3637524  NaT  user 9201  retry 0     time 3.973848819732666  returncode 0   
2399087  NaT  user 9201  retry 0   time 0.05365252494812012  returncode 0   

                                                   command  
1633408  command ['/usr/bin/sbatch', '/tmp/condor_g_scr...  
1941388  command ['/usr/bin/scontrol', 'show', 'job', '...  
4734953  command ['/usr/bin/sbatch', '/tmp/condor_g_scr...  
3637524  command ['/usr/bin/sbatch', '/tmp/condor_g_scr...  
2399087  command ['/usr/bin/sbatch', '/tmp/condor_g_scr...

**Phase 2: Clean and Transform the Data**  
Objectives:  
* Extract job completions from fullsample.csv.  
* Parse CE5 and CE6 logs to identify unresponsive events.  
* Create analysis-ready features (time windows, completion counts, unresponsiveness indicators).  
* Optionally include other features (currently running jobs or resource usage, time-of-day).  
